secrets/certregistry/certregistry.go

```diff
@@ -9,62 +10,114 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
-// CertRegistry object holds TLS certificates to be used to terminate TLS connections
+// Config holds a certificate registry TLS configuration.
+type Config struct {
+	ClientAuth  tls.ClientAuthType
+	Certificate tls.Certificate
+}
+
+// CertRegistry object holds TLS Config to be used to terminate TLS connections
 // ensuring synchronized access to them.
 type CertRegistry struct {
 	mu     sync.Mutex
-	lookup map[string]*tls.Certificate
+	lookup map[string]*tlsConfigWrapper
+
+	// defaultTLSConfig is TLS config to be used as a base config for all host configs.
+	defaultConfig *tls.Config
+}
+
+// tlsConfigWrapper holds the tls.Config and a hash of a host configuration.
+type tlsConfigWrapper struct {
+	config *tls.Config
+	hash   []byte
 }
 
 // NewCertRegistry initializes the certificate registry.
 func NewCertRegistry() *CertRegistry {
-	l := make(map[string]*tls.Certificate)
+	l := make(map[string]*tlsConfigWrapper)
 
 	return &CertRegistry{
-		lookup: l,
+		lookup:        l,
+		defaultConfig: &tls.Config{},
 	}
 }
 
-// Configures certificate for the host if no configuration exists or
-// if certificate is valid (`NotBefore` field) after previously configured certificate.
-func (r *CertRegistry) ConfigureCertificate(host string, cert *tls.Certificate) error {
-	if cert == nil {
-		return fmt.Errorf("cannot configure nil certificate")
+// Configures TLS for the host if no configuration exists or
+// if config certificate is valid (`NotBefore` field) after previously configured certificate.
+func (r *CertRegistry) SetTLSConfig(host string, config *Config) error {
+	if config == nil {
+		return fmt.Errorf("cannot configure nil tls config")
 	}
 	// loading parsed leaf certificate to certificate
-	leaf, err := x509.ParseCertificate(cert.Certificate[0])
+	leaf, err := x509.ParseCertificate(config.Certificate.Certificate[0])
 	if err != nil {
 		return fmt.Errorf("failed parsing leaf certificate: %w", err)
 	}
-	cert.Leaf = leaf
+	config.Certificate.Leaf = leaf
+
+	// Get tls.config and hash from the config
+	tlsConfig, configHash := r.configToTLSConfig(config)
 
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
+	// Check if the config is already set
 	curr, found := r.lookup[host]
-	if found {
-		if cert.Leaf.NotBefore.After(curr.Leaf.NotBefore) {
-			log.Infof("updating certificate in registry - %s", host)
-			r.lookup[host] = cert
-			return nil
-		} else {
+	if found && bytes.Equal(curr.hash, configHash) {
+		return nil
+	}
+
+	if found && !bytes.Equal(curr.hash, configHash) {
+		if !config.Certificate.Leaf.NotBefore.After(curr.config.Certificates[0].Leaf.NotBefore) {
```

Comment:
problem: this causes config to be skipped when separate ingresses have different annotations. We'd need to set the annotation for all hosts regardless of the ingress.